### ISSUE TITLE
Desktop audio capture does not work on Mac and Linux

### ISSUE THREAD

**Issue by mhashmi** [electron/electron#10515]
* Electron version: 1.6.2, 1.6.5, 1.7.6, 2.0.2, 3.x, 4.0.5
* Operating system: Mac OS X Sierra, Linux (Debian Jessie)

### Expected behavior

Given the defined [DesktopCapture API](https://github.com/electron/electron/blob/master/docs/api/desktop-capturer.md) for capturing both audio and video for the entire desktop, we should get a working audio stream on Windows, Linux and Mac.

### Actual behaviour

Electron version 4.0.5: On Mac and Linux, the error callback triggers with the respective messages:

```
getUserMediaError:  DOMException: Could not start audio source // Mac
Failed to query stereo recording // Linux
```

Electron Version 1.7.6: On Mac and Linux, the error callback triggers with the error: 

```
NavigatorUserMediaError
    constraintName: ""
    message: ""
    name: "TrackStartError"
```

Electron Version 1.6.2 and 1.6.5: On Mac and Linux, the video stream is received successfully, albeit, with no audio.

Below are the constraints defined as per the DesktopCapture API: 

```
const constraints = {
  audio: {
    mandatory: {
      chromeMediaSource: 'desktop'
    }
  },
  video: {
    mandatory: {
      chromeMediaSource: 'desktop'
    }
  }
}
```

Please note: on Windows, the audio capture works correctly. 

### How to reproduce

```
git clone https://github.com/mhashmi/electron-desktop-capture
cd electron-desktop-capture
npm install
npm start
```

**Comment by jordanbaucke** [comment]
+1

**Comment by MarshallOfSound** [comment]
@jasonrudolph  Please use GH reactions instead of adding a "+1" comment.

**Comment by jordanbaucke** [comment]
@MarshallOfSound what is a "GH Reaction"? Why is it not mentioned in https://github.com/electron/electron/blob/master/CONTRIBUTING.md#submitting-issues?

**Comment by MarshallOfSound** [comment]
GH reaction == GitHub reaction, you used one just now to thumbs down my comment asking you to use them 🤔

**Comment by jordanbaucke** [comment]
@MarshallOfSound which emoji should I use for confirming the version? since we're getting specific.

**Comment by MarshallOfSound** [comment]
Your initial comment did not contain a version number and therefore could have simply been replaced with a thumbs up on the original issue

**Comment by jasonrudolph** [comment]
> @jasonrudolph Please use GH reactions instead of adding a "+1" comment.

@MarshallOfSound: As someone that got a notification from you for a thread I'm not even participating in, is it helpful for me to say, "Please make sure you're @mentioning the right person."??? 😜

**Comment by MarshallOfSound** [comment]
Haha, sorry Jason.  Been doing this from my phone and didn't check the at mention, my bad 😸

**Comment by leoagle** [comment]
@mhashmi is there other way to fix this bug?

**Comment by mhashmi** [comment]
Guys, a solution has been posted for this issue in bug report #4776, alongside sample code. I've tested it on OS X, and the user who posted it tested it on Linux, so we at least have a workaround now!

**Comment by leoagle** [comment]
@mhashmi thanks for share, but it capture audio from microphone not from system audio.

**Comment by leoagle** [comment]
@mhashmi thanks for share, but it capture audio from microphone not from system or application

**Comment by roy2651** [comment]
@mhashmi that's sound from microphone not from desktop

**Comment by mhashmi** [comment]
Arrgh, I had headphones in my mac and they were lying next to the microphone, which resulted in me thinking the capture was coming through correctly. 

Thank you for pointing it out, and apologies from my end.

**Comment by gio1217** [comment]
same issue

**Comment by peererror** [comment]
same issue

**Comment by kempy007** [comment]
This issue has completely ruined my day/life/idea/business and it's made worse when I try to xvfb it because it seems to think audio is enabled regardless of code?? It totally s**ts a brick with no audio and refuses to stream the video, thus I am totally blocked in my project. 
xwd -display :99 -root -out /tmp/pic # does show that electron has opened browserWindow

**Comment by codebytere** [comment]
We are no longer implementing bugfixes for versions of Electron <= `1.7.x`, so i'm going to close this issue but if it is still persisting in more recent versions of Electron we can certainly reopen it!

**Comment by mhashmi** [comment]
Hi @codebytere. The issue exists in Electron 2.0.2 as well (please see the issue description above and the `Electron versions` line for all versions I tested against). Additionally, the [repository](https://github.com/mhashmi/electron-desktop-capture) replicating the issue targets electron 2.0.2.

Note: I'll find some time to test this again on Electron 2.0.10 as well as Electron 3.x over this week and post an update accordingly.

**Comment by MarshallOfSound** [comment]
> I'll find some time to test this again on Electron 2.0.10 as well as Electron 3.x over this week and post an update accordingly.

Much appreciated, thanks @mhashmi 👍

**Comment by kempy007** [comment]
I can confirm issue persists on 2.0.11 and 3.0

**Comment by mhashmi** [comment]
Sorry about the delay -- as kempy077 pointed out, the issue does indeed exist in both 2.0.11 & 3.0.

**Comment by bryantwells** [comment]
Any updates on this, ideas, alternatives? I'm Having a similar issue. Would love to get desktop audio via an electron app, somehow.

**Comment by EbramTawfik** [comment]
the same issue Mac OS audio :
`navigator.mediaDevices.getUserMedia(
            {
                audio: {
                    mandatory: {
                        chromeMediaSource: 'desktop'
                    }
                },
                video: {
                    mandatory: {
                        chromeMediaSource: 'desktop'
                    }
                }
            }`


I got this :
ERROR:adm_helpers.cc(73)] Failed to query stereo recording.

**Comment by noam-arad** [comment]
I just tested this on electron 4.0.0 and still doesn't work.
The error I get is "NotReadableError: Could not start audio source"

Is there a good workaround at least?

**Comment by brunobgw** [comment]
Does not this have to be with google chomium itself, since even the chromecast itself can not send audio too.
![image](https://user-images.githubusercontent.com/7494476/50595570-1a3c2080-0e88-11e9-8126-de458bdaad60.png)
https://support.google.com/chromecast/answer/3228332?hl=pt-BR&co=GENIE.Platform
* this remark appears when you put the language into Portuguese -Brazil

**Comment by kempy007** [comment]
@noam-arad is the error in v4 different now compared to older releases??

**Comment by codebytere** [comment]
Looking into this today; it seems to DCHECK on master now. I'll try to see how far i get on this and let y'all know.

**Comment by noam-arad** [comment]
> @noam-arad is the error in v4 different now compared to older releases??

I'm really not sure, started using this feature only with Electron 4, I wrote in my comment the error message I get.
I further tried to study this and from my understanding it's an issue that happens all across the MacOS system and not chromium/electronjs specific

**Comment by kempy007** [comment]
Error I currently see with V4.0 with desktop audio capture off so it runs under linux/docker are;

On app startup
```
[130:0106/122507.163532:ERROR:adm_helpers.cc(73)] Failed to query stereo recording.
```

On App opening site with streaming video content
```
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4555:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4555:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4555:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5034:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2565:(snd_pcm_open_noupdate) Unknown PCM default
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4555:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4555:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4555:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5034:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2565:(snd_pcm_open_noupdate) Unknown PCM default
```

running ```aplay some.wav``` works 

and I can see my audio device with lspci

00:05.0 Multimedia audio controller: Intel Corporation 82801AA AC'97 Audio Controller (rev 01)

Trying to extend https://github.com/mhashmi/electron-desktop-capture into a client/server version

**Comment by kempy007** [comment]
BTW I think this error is a red herring since it happens on windows where audio streaming does work from.
 [10364:0106/182117.759:ERROR:adm_helpers.cc(73)] Failed to query stereo recording.

I have thrown this up for testing if it helps
https://github.com/kempy007/Electron-Audio-10515

**Comment by mhashmi** [comment]
@kempy007 The error has indeed changed. I've updated the original report and added 4.0.5 to the list, alongside the error received on the Mac.

Note: I didn't test on Linux as you shared the error -- I've added that to the original report as well. Thanks.

**Comment by tflack** [comment]
Wondering if this is ever going to be fixed or is there some sort of fundamental limitation?

**Comment by bryantwells** [comment]
I ended up spending a decent chunk of time trying to figure this out. This is a limitation put in place by Mac Os X on a system level. Any apps that want to access the system's audio require a signed kernel. From my understanding, this is something that chromium does not provide. 

A little more on the topic here:
https://stackoverflow.com/questions/18443621/mac-os-x-virtual-audio-driver

**Comment by DamienDoumer** [comment]
I've recently had this issue on Mac, though I'm a little bit late in the comments, I have to say I bypassed it using Sound flower. [Here is a detailed blog post about my solution](https://doumer.me/record-system-sound-on-macos-with-javascript-and-soundflower/):

**Comment by advaithm** [comment]
I can also confirm this issue

**Comment by tekktonic** [comment]
Can confirm that there's no reason that this shouldn't work on Linux, does someone mind reopening?

On Linux the process is identical to connecting to a microphone: instead of connecting to the microphone sound source you just connect to the Monitor source of the default output.

**Comment by advaithm** [comment]
It would solve alot of problems with apps downstream. Please fix this issue.

**Comment by upendra-web** [comment]
@tekktonic same here. Tried getting audio sources using `navigator.mediaDevices.enumerateDevices()` API, but couldn't find the `Monitor source of the default output`. 
OS - Manjaro 20.0.3
Electron Version - 8.4.1

**Comment by advaithm** [comment]
@codebytere I believe that this issue needs to be reopened

**Comment by shadsbot** [comment]
The closing message posted only references the Mac side of this issue, and ignored the Linux side.

Can we get some kind of official word that this either won't be followed up for Linux, or have this be reopened and investigated further?

**Comment by ApeironTsuka** [comment]
On the Linux side, if using PulseAudio to pull from the monitor of the default sink is unacceptable when capturing specific apps rather than the whole desktop, you could register a virtual sink and let the end-user manage the routing. Or taking it one step further, register the virtual sink and load module-combine-sink to merge the default sink with the virtual sink. This is likely what the end-user would be doing anyway, so all the they would have to do is temporarily change the app's output sink to the one created by the module.

Personally, if going with the "always pull from the monitor" approach, I'd instead use module-loopback with a virtual sink to do the routing. This way if the end-user has some strange audio setup, they can manually re-route (through their own module-loopback or module-combine-sink or whatever) where they wouldn't be so easily able to if Electron pulls directly from the monitor.

There are many ways this could be implemented on Linux through PulseAudio. People on different audio backends (such as ALSA, or JACK) would still be out of luck, but I'm pretty sure the vast majority of people are using PulseAudio these days.

**Comment by advaithm** [comment]
I tried doing this but its way too complicated if i want to share music while streaming.

**Comment by Kyuunex** [comment]
Can reproduce on Electron/11.0.0-beta.7 on GNU/Linux

pls reopen issue k thx

**Comment by celestial2** [comment]
When i'm using the 6.0.10 version it is working all right.
On any other version the issue is there.
Anyone has a solution for this ?

**Comment by advaithm** [comment]
sadly no which is why we need this to be reopened!!

**Comment by kempy007** [comment]
@celestial2 are you 100% on this?
> **Ubuntu 18.04:**
> 
> When i'm using the 6.0.10 version it is working all right.
> On any other version the issue is there.
> Anyone has a solution for this ?

Can anyone else verify Ubuntu 18.04 + Electron 6.0.10 has working audio??

**Comment by celestial2** [comment]
@kempy007 I've used it so far successfully with that version.
I'm using pulse audio with these 2 modules since i'm running it inside of docker container:
```
pacmd load-module module-null-sink sink_name=Output
pacmd load-module module-virtual-source source_name=Output
```
Tested same environment with versions above 6.0.10, but it doesn't detect the sound there.

**Comment by Kyuunex** [comment]
so, since this is not being reopened, is making a new issue a good idea? idk how these stuff works here

**Comment by Sarthak2319** [comment]
+1 , I also believe that this should be reopened as the Linux side of the issue was ignored when it was closed. Pulseaudio is capable of this , if this is fixed many apps using electron will also be able to use this.

**Comment by advaithm** [comment]
I think someone should make a new splitting Linux and Mac OS as they are 2 different platforms and they should be closed when the issue is resolved

**Comment by advaithm** [comment]
@MarshallOfSound @codebytere would you mind reopening this is issue as on Linux at least there has been a regression which has resulted in audio capture on Linux not working.

**Comment by kempy007** [comment]
@celestial2 do you have a Dockerfile you can share that we could perhaps modify to stream youtube or something with audio using a modified example app as original no longer exist but I found a fork to copy as base https://github.com/kempy007/electron-desktop-capture

**Comment by Kyuunex** [comment]
here's what i use to reproduce this issue [https://github.com/Kyuunex/electron-screen-capture-test](https://github.com/Kyuunex/electron-screen-capture-test) and is still reproducible
i forked and modified from someone's tutorial

**Comment by altermetax** [comment]
Please reopen this issue.

**Comment by TauAkiou** [comment]
@MarshallOfSound @codebytere +1 for reopening this issue. It's extremely frustrating that the MacOS side of this issue came to a resolution, but the Linux side has remained unanswered for over 3 years despite there still being a problem and no functional limitation on the Linux side, unlike on MacOS.

**Comment by TauAkiou** [comment]
I was searching through the Chromium issue tracker, and found this: https://bugs.chromium.org/p/chromium/issues/detail?id=1143761&q=linux%20streaming&can=2

This may be worth keeping an eye on, since it seems related to this issue. It's possible that this issue may see resolution when the Chromium team starts pushing fixes.

**Comment by Kyuunex** [comment]
I wonder if PipeWire can make implementing this much simpler than doing it though PulseAudio

**Comment by Extarys** [comment]
I'm assuming since the above issue [(Chromium bug)](https://bugs.chromium.org/p/chromium/issues/detail?id=1143761&q=linux%20streaming&can=2) is still open that Linux audio capture is still not implemented for Electrons apps?

Just gently pinging this as we are now in february :smile:

**Comment by owarie** [comment]
This issue is still occurring on Linux, pls reopen

**Comment by Fiser12** [comment]
I have this same problem currently, I found the error on Linux and on Mac OS, pls reopen it

**Comment by procopio420** [comment]
im facing the same issue right now :( pls reopen it

**Comment by arthur4ires** [comment]
I am also facing the problem, please re-open: "(

**Comment by Extarys** [comment]
@codebytere I'm not sure which version Discord uses but it seems that this might not be fixed yet.

**Comment by 3N4N** [comment]
@codebytere I know you guys have a lot of issues at hand and too little manpower, but leaving a bug in Linux for one of the most important features unfixed for four years with no communication is detrimental to the Linux community. Please check into it once again.

**Comment by AmarBouchibane** [comment]
I also had the problem in Ubuntu (video with audio stream of a BrowserWindow) and solved it in our app with the help of "paclient" (https://github.com/mscdex/paclient).

```
// Move  the  specified  recording  stream  (identified by its numerical index) 
// to the specified source (identified by its symbolic name or numerical index).
// e.g. "pacmd move-source-output ID alsa_output.pci-0000_02_02.0.analog-stereo.monitor"
```

Then the corresponding deviceId can be used in the audio constraints and mixed with the video stream to get a video-with-audio stream.
But a simpler solution would of course be better.

**Comment by Shubham-Kumar-2000** [comment]
Still facing this issue on linux. Please reopen.

**Comment by Adi-11** [comment]
I'm facing the same issue right now, pls reopen it

**Comment by cybe42** [comment]
The same issue is still in Linux. Please reopen.

**Comment by Spark61** [comment]
I have also the same issue

**Comment by weedsboy** [comment]
same issue on linux

ubuntu 22.04
electron 9.4.0

**Comment by sniff122** [comment]
ngl this is one of the many reasons why i dont like electron, this issue has existed since september 2017, was fixed for mac and basically ignored for linux, absolutely terrible

**Comment by ApeironTsuka** [comment]
> ngl this is one of the many reasons why i dont like electron, this issue has existed since september 2017, was fixed for mac and basically ignored for linux, absolutely terrible. This issue is not resolved so it shouldnt have been closed

@sniff122 
To be fair to Electron, this issue is blocked by Chromium itself. [And they don't seem to be in a hurry to ever fix it.](https://bugs.chromium.org/p/chromium/issues/detail?id=1143761) It should theoretically be possible for Electron to roll their own fix in, but that's largely outside the scope of the project.

A kludgy sort of "fix" (at least for PulseAudio-based systems) would be to expose a virtual sink when capturing and leave it as an exercise for the user to get the application/system audio into it (likely via a combination of module-loopback and module-combine-sink), but that's very user-unfriendly.

**Comment by tekktonic** [comment]
As you can see in the issue linked, the chromium team considers it "won't fix" however the fix is trivial and I believe a patch to not filter out the monitor devices is provided in the report. There's no reason this patch couldn't be applied by the Electron team.

**Comment by Hrishabh17** [comment]
> 

Hi, I am also stuck with the system audio not recording in Linux, could you show me how you achieved it? Like where have you added this code? Thank you, its really urgent.

**Comment by Hrishabh17** [comment]
> I also had the problem in Ubuntu (video with audio stream of a BrowserWindow) and solved it in our app with the help of "paclient" (https://github.com/mscdex/paclient).
> 
> ```
> // Move  the  specified  recording  stream  (identified by its numerical index) 
> // to the specified source (identified by its symbolic name or numerical index).
> // e.g. "pacmd move-source-output ID alsa_output.pci-0000_02_02.0.analog-stereo.monitor"
> ```
> 
> Then the corresponding deviceId can be used in the audio constraints and mixed with the video stream to get a video-with-audio stream. But a simpler solution would of course be better.

@AmarBouchibane Hi, I am also stuck with the system audio not recording in Linux, could you show me how you achieved it? Like where have you added this code? Thank you, it's really urgent.

**Comment by Hrishabh17** [comment]
> Is this fixed? I still have the same error.

Yes, for Linux I used pulse audio to redirect the system audio to my application and thence use it. If you need the command to create the virtual sink and source I can share it. You can execute the command in the pre-loader of the electron app. For mac, there is an alternative (sound flower). I was mostly concerned with linux and windows so never bothered about mac.

**Comment by KonstantinBel** [comment]
> > Is this fixed? I still have the same error.
> 
> Yes, for Linux I used pulse audio to redirect the system audio to my application and thence use it. If you need the command to create the virtual sink and source I can share it. You can execute the command in the pre-loader of the electron app. For mac, there is an alternative (sound flower). I was mostly concerned with linux and windows so never bothered about mac.

I'd like to see your solution, I have the same problem, but don't have an experience with Linux to understand how to use PulsAudio.

**Comment by kerberizer** [comment]
Some possibly good news. Apparently, work on the [upstream issue](https://bugs.chromium.org/p/chromium/issues/detail?id=1143761) in Chromium has started on Aug 18, 2023.

**Comment by Hrishabh17** [comment]
> > > Is this fixed? I still have the same error.
> > 
> > 
> > Yes, for Linux I used pulse audio to redirect the system audio to my application and thence use it. If you need the command to create the virtual sink and source I can share it. You can execute the command in the pre-loader of the electron app. For mac, there is an alternative (sound flower). I was mostly concerned with linux and windows so never bothered about mac.
> 
> I'd like to see your solution, I have the same problem, but don't have an experience with Linux to understand how to use PulsAudio.

sure, what I did is:

I used spawn to run the commands as follow:

***********************************************************************************************************
Step 1:

const monitorProcess = spawn('pacmd list-sources | grep -B 1 analog-stereo.monitor | grep name:', {
  shell: true
});

this actually checks if there is already a monitor or not (simple way we need to create this monitor if its not present, there is one default monitor but its of no use, so to route the system audio we need to create this, I'll reference it as user monitor hence forth to explain other steps"

****************************************************************************************************************
Step 2: 

monitorProcess.stdout.on('data', function (data) {
  const output = data.toString();
  const l = output.split('\n')
  if(l.length <= 2){
    const regex = /name:\s*<(.+)>/i; 
    const match = output.match(regex);
    let monitorName
    if (match && match.length > 1) {
      monitorName = match[1];
    }
    const moduleGenerate = spawn(`pactl load-module module-remap-source master=${monitorName} source_properties=device.description=Computer-sound`, {
      shell:true
    })
  
    moduleGenerate.stdout.on('end', ()=>{
    })
  }
});

This simply checks the if the user monitor already created (monitors refresh/reset after a system restart, if its already present then no need to create it). (to get more idea about this simply run the commands in step1 followed by step2 and find it your self, do it for 5-6 times and you will be comfortable with what we are doing)

**************************************************************************************************************************************

Step 3:

Once the monitor is created, we can use this in the frontend (webRTC call) and it will be listed as one of the sources. This source will acts a way to get the system audio.

**************************************************************************************************************************************

Step 4:

Once this is done, you can take the microphone audio source as well (two webRTC calls) then you'll have to combine these two sources (if you need help regarding this I would be happy to help, this too can be done with vanilla javascript without any packages)

**Comment by GitHubCloud** [comment]
> I've recently had this issue on Mac, though I'm a little bit late in the comments, I have to say I bypassed it using Sound flower. [Here is a detailed blog post about my solution](https://doumer.me/record-system-sound-on-macos-with-javascript-and-soundflower/): This include every code snippet to record system audio on MacOS and even video and then combine both streams to make audio and video together.

I know it's been quite a time, but were you able to build a package with the SoundFlower solution?

**Comment by benkingcode** [comment]
Chromium has apparently fixed this now: https://issues.chromium.org/issues/40155218

Will Electron be gaining this functionality?
